hedera-node/hedera-app/src/main/java/com/hedera/node/app/util/HashTreeManager.java

```diff
@@ -0,0 +1,183 @@
+/*
+ * Copyright (C) 2024 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hedera.node.app.util;
+
+import com.hedera.pbj.runtime.Codec;
+import com.hedera.pbj.runtime.io.buffer.Bytes;
+import java.nio.charset.StandardCharsets;
+import java.security.MessageDigest;
+import java.util.ArrayList;
+import java.util.List;
+
+public class HashTreeManager<T> {
+
+    public HashTreeManager(Codec<T> codec) {
+        this.codec = codec;
+    }
+
+    private final Codec<T> codec;
+    List<Bytes> inputsCompleteSubtreeHashes;
+    List<Bytes> outputsCompleteSubtreeHashes;
+    private List<Bytes> hashList = new ArrayList<>();
+
+    private enum Alternator {
+        LEFT,
+        RIGHT
+    }
+
+    Alternator currentAlternate = Alternator.LEFT;
+
+    public void processNodes(List<T> elements) {
```

Comment:
Perhaps we should have another method which accepts as an argument a single element since I am not sure how we will be calling this (one block item at a time most likely).

Reply:
Agreed and added. Thanks @derektriley

Reply:
Definitely. I dunno that we'll ever want to pass in a collection at all, since we want to be hashing elements as quickly as possible, independent of the handle workflow. Passing in a collection with size > 1 means that at least one element is sitting in that list waiting for the construction of a separate element, which seems to violate (even if only slightly) the goal of hashing as soon as the block item becomes available.

Reply:
Makes sense. Have removed the addElements method as well to reflect 👍